ext/php8/handlers_curl.c

```diff
@@ -89,6 +100,110 @@ static void dd_init_headers_arg(zval *arg, zval *ch) {
 
 static void dd_free_headers_arg(zval *arg) { zend_array_destroy(Z_ARRVAL_P(arg)); }
 
+static void dd_inject_distributed_tracing_headers(zval *ch) {
+    if (dd_inject_func == NULL) {
+        dd_inject_func = zend_string_init(ZEND_STRL("ddtrace\\bridge\\curl_inject_distributed_headers"), 0);
+    }
+    if (zend_hash_exists(EG(function_table), dd_inject_func)) {
+        zend_function **fn_proxy = &dd_curl_inject_fn_proxy;
+        zval retval = {.u1.type_info = IS_UNDEF};
+
+        zval headers;
+        dd_init_headers_arg(&headers, ch);
+
+        ddtrace_sandbox_backup backup = ddtrace_sandbox_begin();
+        dd_should_save_headers = false;  // Don't save our own HTTP headers
+        // Arg 0: CurlHandle $ch
+        // Arg 1: mixed $value (array of headers)
+        if (ddtrace_call_function(fn_proxy, ZSTR_VAL(dd_inject_func), ZSTR_LEN(dd_inject_func), &retval, 2, ch,
+                                  &headers) == SUCCESS) {
+            zval_ptr_dtor(&retval);
+        } else {
+            ddtrace_log_debug("Could not inject distributed tracing headers");
+        }
+        dd_should_save_headers = true;
+        ddtrace_sandbox_end(&backup);
+
+        dd_free_headers_arg(&headers);
+    }
+}
+
+static void dd_zo_ch_dtor(void *ch) {
+    zend_object *zo_ch = *((zend_object **)ch);
+    GC_DELREF(zo_ch);
+}
+
+static void dd_update_mh_ch_map_cache(zval *mh, HashTable *ch_map) {
+    dd_mh_ch_map_cache_id = Z_OBJ_HANDLE_P(mh);
+    dd_mh_ch_map_cache = ch_map;
+}
+
+static void dd_mh_map_add_ch(zval *mh, zval *ch) {
+    HashTable *ch_map = NULL;
+    zend_object *zo_ch = Z_OBJ_P(ch);
+
+    if (!dd_mh_ch_map) {
+        ALLOC_HASHTABLE(dd_mh_ch_map);
+        zend_hash_init(dd_mh_ch_map, 8, NULL, (dtor_func_t)dd_ht_dtor, 0);
+    } else {
+        ch_map = zend_hash_index_find_ptr(dd_mh_ch_map, Z_OBJ_HANDLE_P(mh));
+    }
+
+    if (!ch_map) {
+        ALLOC_HASHTABLE(ch_map);
+        zend_hash_init(ch_map, 8, NULL, (dtor_func_t)dd_zo_ch_dtor, 0);
+        zend_hash_index_update_ptr(dd_mh_ch_map, Z_OBJ_HANDLE_P(mh), ch_map);
+    }
+
+    GC_ADDREF(zo_ch);
+    zend_hash_index_update_ptr(ch_map, Z_OBJ_HANDLE_P(ch), zo_ch);
+
+    dd_update_mh_ch_map_cache(mh, ch_map);
+}
+
+static void dd_mh_map_remove_ch(zval *mh, zval *ch) {
+    HashTable *ch_map = NULL;
+
+    if (dd_mh_ch_map) {
+        ch_map = zend_hash_index_find_ptr(dd_mh_ch_map, Z_OBJ_HANDLE_P(mh));
+        if (ch_map) {
+            zend_hash_index_del(ch_map, Z_OBJ_HANDLE_P(ch));
+        }
+        dd_update_mh_ch_map_cache(mh, ch_map);
+    }
```

Comment:
Can you add a comment on this? Seems weird to find the index to delete; so explain it in the comment.